precompiles/dapps-staking/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+            era, reward, staked, gas_used
+        );
+
+        let reward = TryInto::<u128>::try_into(reward).unwrap_or(0);
+        let mut output = utils::argument_from_u128(reward);
+
+        let staked = TryInto::<u128>::try_into(staked).unwrap_or(0);
+        let mut staked_vec = utils::argument_from_u128(staked);
+        output.append(&mut staked_vec);
+
+        Ok(PrecompileOutput {
+            exit_status: ExitSucceed::Returned,
+            cost: gas_used,
+            output,
+            logs: Default::default(),
+        })
+    }
+
+    /// Fetch registered contract from RegisteredDevelopers storage map
+    fn registered_contract(input: EvmInArg) -> Result<PrecompileOutput, ExitError> {
+        println!("--- precompile registered_contract() {:?}", input.len());
+        let developer_h160 = input.to_h160(1);
+        let developer = R::AddressMapping::into_account_id(developer_h160);
+        println!("--- precompile developer_h160 {:?}", developer_h160);
+        println!("--- precompile developer public key {:?}", developer);
+
+        let smart_contract = pallet_dapps_staking::RegisteredDevelopers::<R>::get(&developer);
+        let gas_used = R::GasWeightMapping::weight_to_gas(R::DbWeight::get().read);
+
+        println!(
+            "--- precompile developer {:?}, contract {:?}",
+            developer, smart_contract
+        );
+        let output = argument_from_h160_vec(smart_contract.unwrap_or_default().encode());
+
+        Ok(PrecompileOutput {
+            exit_status: ExitSucceed::Returned,
+            cost: gas_used,
+            output,
+            logs: Default::default(),
+        })
+    }
+
+    /// Register contract with the dapp-staking pallet
+    fn register(input: EvmInArg) -> Result<R::Call, ExitError> {
+        println!("--- precompile register() {:?}", input.len());
+        input.expecting_arguments(1).map_err(|e| exit_error(e))?;
+        // parse contract's address
+        let contract_h160 = input.to_h160(1);
+        // println!("contract_h160 {:?}", contract_h160);
+
+        let contract_id = Self::decode_smart_contract(contract_h160)?;
+
+        Ok(pallet_dapps_staking::Call::<R>::register { contract_id }.into())
+    }
+
+    /// Lock up and stake balance of the origin account.
+    fn bond_and_stake(input: EvmInArg) -> Result<R::Call, ExitError> {
+        input.expecting_arguments(2).map_err(|e| exit_error(e))?;
+
+        // parse contract's address
+        let contract_h160 = input.to_h160(1);
+        // println!("contract_h160 {:?}", contract_h160);
+        let contract_id = Self::decode_smart_contract(contract_h160)?;
+
+        // parse balance to be staked
+        let value = input.to_u256(2).low_u128().saturated_into();
+        println!("--- precompile bond value {:?}", value);
+
+        Ok(pallet_dapps_staking::Call::<R>::bond_and_stake { contract_id, value }.into())
+    }
+    /// Helper method to decode type SmartContract enum
+    fn decode_smart_contract(
```

Comment:
We discussed this so perhaps comment is redundant - SmartContract should be a common type to avoid code duplication.

Reply:
I prefer to do this after this PR